### PR TITLE
[v6r20] new fts3 only rebase

### PR DESCRIPTION
**Pull request by chaen**
Replaces https://github.com/DIRACGrid/DIRAC/pull/2826

### CODE REVIEW

**Comment by schr476** [comment]
We plan to test on new BelleDIRAC branch now that we delivered the latest DDM for BelleDIRAC Monte Carlo 9 (starts July 5th)

File: DataManagementSystem/Agent/FTS3Agent.py [hunk 1]
```diff
@@ -0,0 +1,420 @@
...
+        log.error( "Error updating file fts status", "%s, %s" % ( ftsJob.ftsGUID, res ) )
+        return ftsJob, res
+  
+
+      upDict = {ftsJob.jobID :  {'status' : ftsJob.status,
+                                                              'error' : ftsJob.error,
+                                                              'completeness' : ftsJob.completeness,
+                                                              'operationID' : ftsJob.operationID,
+                                                              }
+                                          }
```
Comment (fstagni) [inline DataManagementSystem/Agent/FTS3Agent.py]: Weird indentation

File: DataManagementSystem/Agent/FTS3Agent.py [hunk 2]
```diff
@@ -0,0 +1,420 @@
...
+      res = self.fts3db.updateJobStatus( upDict )
+
+      return ftsJob, res
+
+    except Exception as e:
+      return ftsJob, S_ERROR( "Exception %s" % repr( e ) )
```
Comment (fstagni) [inline DataManagementSystem/Agent/FTS3Agent.py]: DErrno?

File: DataManagementSystem/Agent/FTS3Agent.py [hunk 3]
```diff
@@ -0,0 +1,420 @@
...
+    return S_OK()
+
+      
+
+
+
+        
+
+
+
```
Comment (fstagni) [inline DataManagementSystem/Agent/FTS3Agent.py]: You pushed "Enter" too many times :)

File: DataManagementSystem/Agent/RequestOperations/ReplicateAndRegister.py [hunk 4]
```diff
@@ -310,18 +320,26 @@ def ftsTransfer( self ):
...
     if not res['OK']:
       return res
     else:
-      filesToScheduleList = res['Value']
+      filesToSchedule = res['Value']
+
+      for lfn in filesToSchedule:
+        filesToScheduleList.append( ( filesToSchedule[lfn][0].toJSON()['Value'],
+                              toSchedule[lfn][1],
+                              toSchedule[lfn][2] ) )
```
Comment (fstagni) [inline DataManagementSystem/Agent/RequestOperations/ReplicateAndRegister.py]: wrong indentation

File: DataManagementSystem/Client/FTS3Client.py [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+    if not res['OK']:
+      return res
+
+    operationsJSON = res['Value']
+
+    try:
+      operations = json.loads( operationsJSON, cls = FTS3JSONDecoder )
+      return S_OK( operations )
+    except Exception as e:
+      return S_ERROR( "Exception when decoding the non finished operations json %s" % e )
```
Comment (fstagni) [inline DataManagementSystem/Client/FTS3Client.py]: DError?

File: DataManagementSystem/DB/FTS3DB.py [hunk 6]
```diff
@@ -0,0 +1,456 @@
+__RCSID__ = "$Id $"
+
+# pylint: disable=no-member
```
Comment (fstagni) [inline DataManagementSystem/DB/FTS3DB.py]: For what?

File: DataManagementSystem/Client/FTS3Job.py [hunk 7]
```diff
@@ -0,0 +1,506 @@
+import datetime
```
Comment (fstagni) [inline DataManagementSystem/Client/FTS3Job.py]: missing __RCSID__

File: DataManagementSystem/Agent/FTS3Agent.py [hunk 8]
```diff
@@ -0,0 +1,420 @@
+"""
```
Comment (fstagni) [inline DataManagementSystem/Agent/FTS3Agent.py]: missing __RCSID__

File: DataManagementSystem/DB/FTS3DB.py [hunk 9]
```diff
@@ -0,0 +1,456 @@
...
+
+      # Merge it in case it already is in the DB
+      operation = session.merge( operation )
+      session.add( operation )
+      session.commit()
+      session.expunge_all()
+
+      return S_OK( operation.operationID )
+
+    except SQLAlchemyError, e:
```
Comment (fstagni) [inline DataManagementSystem/DB/FTS3DB.py]: `as e` (in several places)

File: DataManagementSystem/DB/FTS3DB.py [hunk 10]
```diff
@@ -0,0 +1,456 @@
...
+      session.commit()
+
+      ###################################
+      session.expunge_all()
+      return S_OK( operation )
+
+    except NoResultFound, e:
+      return S_ERROR( "No FTS3Operation with id %s" % operationID )
+    except SQLAlchemyError, e:
+      return S_ERROR( "getOperation: unexpected exception : %s" % e )
```
Comment (fstagni) [inline DataManagementSystem/DB/FTS3DB.py]: DError, also 2 lines above

File: DataManagementSystem/Service/FTS3ManagerHandler.py [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+.. module: FTS3ManagerHandler
+  :synopsis: handler for FTS3DB using DISET
+
+Service handler for FT3SDB using DISET
+"""
+
+__RCSID__ = "$Id$"
+
+import json
+from types import DictType, IntType, LongType, ListType, StringTypes, NoneType
```
Comment (fstagni) [inline DataManagementSystem/Service/FTS3ManagerHandler.py]: Don't. Just use base types

File: docs/source/AdministratorGuide/Systems/DataManagement/fts3.rst [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+    * Since their might be several possible source SE, we need to pick one only. The choice is to select the SE where there is the most files of the operation present. This increases the likely hood to pick a good old Tier1
+    * Divide all that according to the maximum number of files we want per job
+
+Once the FTS jobs have been executed, and all the operation is completed, the callback takes place. The callback consists in fetching the RMS request which submitted the FTS3Operation, update the status of the RMS files, and insert a Registration Operation.
+Note that since the multiple targets are grouped in a single RMS operation, failing to transfer one file t one destination will result in the failure of the Operation. However, there is one Registration operation per target, and hence correctly transferred files will be registered.
+
+FTS3StagingOperation
+====================
+
+**CAUTION: still in development, not meant to be used **
```
Comment (fstagni) [inline docs/source/AdministratorGuide/Systems/DataManagement/fts3.rst]: Is it?

File: requirements.txt [hunk 13]
```diff
@@ -33,3 +33,4 @@ six
 sqlalchemy>=1.0.9
 stomp.py>=3.1.5
 suds-jurko>=0.6
+git+https://gitlab.cern.ch/fts/fts-rest.git@master
```
Comment (fstagni) [inline requirements.txt]: I think this is not needed now (it is at line 2 here)

**Comment by chaen** [comment]
I am not going to spend one more second on this PR unless NOW is really the time it is going to be merged.

**Comment by atsareg** [comment]
I have not understood the last comment by @chaen. The PR is marked for v6r20. It can be merged now after the v6r19 is prepared into integration for the next release. This PR has several unanswered comments - are you going to react to them ?

**Comment by chaen** [comment]
This PR was originally targeted for v6r15 I believe, and I have rebased and retested many times, because "finally, not for this release". I will not do this exercise anymore, nor will I react to the comments, unless I am sure it will not be postponed again. If so, I am happy to finalize it.

**Comment by atsareg** [comment]
I do not remember why this PR was not targeted for v6r18,19. There was a reason. Nevertheless, if this is no more the WIP grade PR, I suggest to merge it immediately into the current integration and make it the main subject of the next release. It can be the only main new feature and we can make the new release very quickly. If you agree, please, review it (for the last time). There are few comments from Federico that need your attention.

**Comment by chaen** [comment]
The reason was reshuffling priorities. 
OK, doing it now.

**Comment by chaen** [comment]
I answer the comment inlines, but I make a new PR

File: DataManagementSystem/Agent/FTS3Agent.py (hunk 8 above)
Comment (chaen) [reply DataManagementSystem/Agent/FTS3Agent.py]: Done

File: DataManagementSystem/Agent/FTS3Agent.py (hunk 1 above)
Comment (chaen) [reply DataManagementSystem/Agent/FTS3Agent.py]: Reformat using Pep8

File: DataManagementSystem/Agent/FTS3Agent.py (hunk 2 above)
Comment (chaen) [reply DataManagementSystem/Agent/FTS3Agent.py]: Delicate for such a general exception.
I added DErrno where possible now

File: DataManagementSystem/Agent/FTS3Agent.py (hunk 3 above)
Comment (chaen) [reply DataManagementSystem/Agent/FTS3Agent.py]: :-)

File: DataManagementSystem/DB/FTS3DB.py (hunk 6 above)
Comment (chaen) [reply DataManagementSystem/DB/FTS3DB.py]: All the class attributes generated by SQLAlchemy

Comment (chaen) [reply DataManagementSystem/DB/FTS3DB.py]: I added a comment

File: DataManagementSystem/DB/FTS3DB.py (hunk 9 above)
Comment (chaen) [reply DataManagementSystem/DB/FTS3DB.py]: Thanks, fixed

File: DataManagementSystem/Service/FTS3ManagerHandler.py (hunk 11 above)
Comment (chaen) [reply DataManagementSystem/Service/FTS3ManagerHandler.py]: Fixed

File: docs/source/AdministratorGuide/Systems/DataManagement/fts3.rst (hunk 12 above)
Comment (chaen) [reply docs/source/AdministratorGuide/Systems/DataManagement/fts3.rst]: Yep, because not tested

File: requirements.txt (hunk 13 above)
Comment (chaen) [reply requirements.txt]: Yep, removed

**Pull request closed without merging**